Add Room Audio Control to Control4 Integration

nalin29 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add Media Player Device to Control4 Integration. This change allows users to operate their room sound system by changing sources, changing volume, muting and turning off the room.

Dependency Diff:

https://github.com/lawtancool/pyControl4/compare/v1.0.0b6...v1.1.0

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/31910

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

lawtancool commented:
@nalin29 Thanks for this! Is the only reason for the forked dependency this PR? https://github.com/lawtancool/pyControl4/pull/20

If so, we could work to get that merged and avoid the dependency change.

lawtancool commented on homeassistant/components/control4/director_utils.py:
```diff
@@ -38,6 +40,32 @@ async def director_update_data(
     return {key["id"]: key for key in data}
 
 
+async def update_variables_for_entity(
```
Is there a reason that we couldn't use this function for the light component as well? This seems like it could possibly replace `director_update_data`, which would be preferable to avoid duplication

nalin29 replied:
Yes, I believe so. For my current work I just wanted to focus on the media player and make no changes to other portions of the integration in case of breakage.

nalin29 replied:
Done, I tested it on my own lights which include dimmable and non-dimmable.

lawtancool commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,372 @@
+
+@attr.s
+class RoomSource:
+    """Error during condition evaluation."""
```
What does "Error during condition evaluation." mean?

nalin29 replied:
Fixed Comment

lawtancool commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,372 @@
+            coordinator,
+            name,
+            idx,
+            device_name=f"Control4 Room - {name}",
```
Since rooms don't have real device names, manufacturers, or models, can we just pass `None` into the super constructor instead?

nalin29 replied:
Possibly, we could pass None, but I am not sure if this will cause an error or make it appear oddly in the HA UI. This is my first time working with HA,

lawtancool replied:
Could you test using `None`? We should avoid make up fake device names when they don't actually exist.

nalin29 replied:
Done, works fine but sets the various properties to <unknown>which is inconsequential.

lawtancool commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,372 @@
+
+        await self.coordinator.async_request_refresh()
+
+    def turn_on(self):
```
There is a `async_turn_off` function that actually calls a Control4 API call to turn off the room - is it not possible to tell Control4 to turn the room on?

nalin29 replied:
No, the rooms are always "on" in a sense. They become powered when we select a source for the room thus we have this dummy turn on. The speakers can be turned "off" in the traditional sense via the API call.

nalin29 commented:
> @nalin29 Thanks for this! Is the only reason for the forked dependency this PR? [lawtancool/pyControl4#20](https://github.com/lawtancool/pyControl4/pull/20)
> 
> If so, we can work to get that merged and avoid the dependency change.

Yep, its only that. If that gets merged then we do not need the forked version of pyControl4.

nalin29 commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,372 @@
+            device_model=None,
+            device_id=idx,
+        )
+        self._attr_device_class = MediaPlayerDeviceClass.SPEAKER
```
I tried to get Video working but unfortunately there was no way that I could find in any of the room variables of determining correctly what source was playing so I wanted to narrow the source to audio control.

nalin29 replied:
I can remove the device class all it really does is tell control4 what icons to use but I do think its good to have it specified.

nalin29 replied:
So I can check the "PLAYING_AUDIO_DEVICE" variable to get the device id for playing audio source. There is also a "CURRENT_AUDIO_DEVICE" variable but this seems to always point to the actual audio "device" like an AMP, or in my case Audio Matrix digital audio output. Similarly, we have the CURRENT_VIDEO_DEVICE and CURRENT_VIDEO_AUDIO_DEVICE. There is also a "CURRENT_SELECTED_DEVICE" which seems to alternate between the CURRENT variables depending on what was chosen last. We can have both  CURRENT_VIDEO_DEVICE and CURRENT_AUDIO_DEVICE be populated with different devices at the same time. Thus, the only thing that I can confirm is that PLAYING_AUDIO_DEVICE is always accurate as it provides the direct source or device that is playing audio.

nalin29 replied:
Maybe I could first check the CURRENT_SELECTED_DEVICE and if PLAYING_AUDIO_DEVICE is populated let that be the current source. But if it is not then display the CURRENT_SELECTED_DEVICE as the source.

nalin29 replied:
Done. I set the Media Player Class to be TV if any video sources are discovered and Media type to be video if playing. The current selected source is determined by the selected device and is chosen to be playing audio device if it is populated and the current selected device is the current audio device.

nalin29 replied:
Tested and Works like a charm

nalin29 replied:
I used the MEDIA_INFO which according to C4 community is what the app uses to display current media data. This lets me get the device, media source device if the source is a media server as well as the title.

nalin29 replied:
There is not much more I can do in terms of detecting audio sources and types. So this is pretty much final.

nalin29 commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,372 @@
+            | MediaPlayerEntityFeature.TURN_OFF
+            | MediaPlayerEntityFeature.TURN_ON
+            | MediaPlayerEntityFeature.SELECT_SOURCE
+            | MediaPlayerEntityFeature.GROUPING
```
No, I looked at trying but I couldn't find any way to do it with the current known APIs

nalin29 commented:
Note I squashed all previous commits into one

nalin29 commented:
@lawtancool I added a new pull request for pycontrol4 with the suggested changes and squashed all commits: https://github.com/lawtancool/pyControl4/pull/21

nalin29 commented:
Updated the dependency fixed the code and tested

nalin29 commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,413 @@
+CONTROL4_PLAYING = "PLAYING"
+CONTROL4_PAUSED = "PAUSED"
+CONTROL4_STOPPED = "STOPPED"
+CONTROL4_MEDIA_INFO = "CURRENT MEDIA INFO"
```
No this is not a mistake. For some, probably dumb reason, in control4 this room variable has spaces in it.

nalin29 commented:
Just checking what are the next steps since you approved the changes. Do we need another approval before it can be merged?

Xorso commented:
Would love to see this get merged.

lorddaren reviewed (approved):
I know @lawtancool can be busy. It looks like everything is resolved. I have helped implement the socket.io portion of the control4 library so I am going to mark this as approved.

EDIT:  Although it looks like it was already approved by @lawtancool. I don't have write access so maybe we are just waiting for him or another from the home-assistant team to merge the commit?

lawtancool commented:
Hi @Xorso , good to see you again!

Since I've approved this we're just waiting on someone from Home Assistant to review and approve, I'm not sure how long that will take since they have a lot of PRs to go through.

epenet commented on homeassistant/components/control4/manifest.json:
```diff
@@ -6,7 +6,7 @@
   "iot_class": "local_polling",
   "loggers": ["pyControl4"],
-  "requirements": ["pyControl4==0.0.6"],
+  "requirements": ["pyControl4==1.1.0"],
```
I suggest that you move this to a preliminary PR.

nalin29 replied:
got it

nalin29 replied:
https://github.com/home-assistant/core/pull/90115

epenet commented on homeassistant/components/control4/light.py:
```diff
@@ -185,13 +187,15 @@ def create_api_object(self):
     def is_on(self):
         """Return whether this light is on or off."""
-        return self.coordinator.data[self._idx]["value"] > 0
+        if self._is_dimmer:
+            return self.coordinator.data[self._idx][CONTROL4_DIMMER_VAR] > 0
+        return self.coordinator.data[self._idx][CONTROL4_NON_DIMMER_VAR] > 0
```
This looks like an unrelated bugfix. It should be a separate PR.

nalin29 replied:
This is a refactor to make use the new version of the helper function "update variables for entity"  to prevent unnecessary redundant helper functions. I will include it in that new preliminary PR along with a version bump and the helper function change

epenet replied:
Please consider instead making multiple preliminary PR.
Having one big preliminary PR would defy the point...

nalin29 replied:
Then what should be included in multiple preliminary PRs? The first PR just has the dependency bump and const changes and the second PR contains the new function and changes in the light platform.

epenet replied:
Something like that - see https://developers.home-assistant.io/docs/review-process#creating-the-perfect-pr

nalin29 replied:
For the second preliminary PR should I wait for the first PR to be merged then introduce it

epenet replied:
If it depends on the first one - please wait until it is merged.
If it is independant, you can create it now.

nalin29 replied:
Resolved by https://github.com/home-assistant/core/pull/90234

nalin29 commented on homeassistant/components/control4/director_utils.py:
```diff
@@ -21,21 +23,30 @@
-async def director_update_data(
-    hass: HomeAssistant, entry: ConfigEntry, var: str
-) -> dict:
+async def update_variables_for_entity(
```
got it

nalin29 commented:
First Preliminary PR involving the version bump and any changes required for the version change.

https://github.com/home-assistant/core/pull/90115

nalin29 commented:
Next Small preliminary PR at https://github.com/home-assistant/core/pull/90187

nalin29 commented:
Second to last PR request. Includes changes to helper function and light platform. Everything is in place to add the new media player platform https://github.com/home-assistant/core/pull/90234

nalin29 commented:
Just checking in are we still working on getting this pulled since it's been a while?

nalin29 commented:
@epenet When can we start to finishing this pull request. The last media module is at https://github.com/home-assistant/core/pull/90501 . If i can get this approved we should be done

Xorso commented:
@nalin29 It would be nice to see this. Anyway you can make a PR request also for the HACS branch so I can test it out?

nalin29 commented:
> @nalin29 It would be nice to see this. Anyway you can make a PR request also for the HACS branch so I can test it out?

You can use https://github.com/home-assistant/core/pull/90501. 

I also have a custom integration repo at https://github.com/nalin29/control4RoomMedia

nalin29 committed:
Add control4 room based media player (#13)

nalin29 commented:
Merged changes from mediacontrol to make this the final pull request making it easier for the doc pull request to be accepted. 

Soon, I will be unable to access the control4 system so if this pull requested does not get accepted or worked on soon this feature will never make it to HA.

nalin29 commented:
> Thanks for splitting out/sorting out all the preliminary PRs.
> 
> Sadly for various reasons I have not been in a position to review PRs recently and won't be able to do so for a few months still.

Is there another HA dev that could be assigned? This should be close to done so if someone else has the time we can get this pulled.

edenhaus commented:
Is #90501 a duplicate of this PR?

At least @nalin29 your are the author of both and the changes look similar

nalin29 commented:
@edenhaus yes I moved the changes over to here since this was the original PR which required multiple preliminary PRs and is attached to documentation changes should it be merged

nalin29 committed:
Merge branch 'dev' into dev

nalin29 commented:
@edenhaus rebased to fix conflict

emontnemery commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,406 @@
+    VIDEO = 2
+
+
+@attr.s
+class RoomSource:
+    """Room Source Data."""
+
+    source_type: set[_SourceType] = attr.ib()
+    id: int = attr.ib()
+    name: str = attr.ib()
```
There's no need to us `attr` here, please make it a dataclass

emontnemery commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,406 @@
+
+    entry_data = hass.data[DOMAIN][entry.entry_id]
+    scan_interval = entry_data[CONF_SCAN_INTERVAL]
+    _LOGGER.debug(
+        "Scan interval = %s",
+        scan_interval,
+    )
```
```suggestion
    _LOGGER.debug("Scan interval = %s", scan_interval)
```

emontnemery commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,406 @@
+    )
+
+    # Fetch initial data so we have data when entities subscribe
+    await coordinator.async_refresh()
```
What happens if this raises `UpdateFailed` or some errors, will the media players still work after a successful refresh?
The same potential issue exists in the light platform.

If this is a problem, maybe the coordinators (this one and the ones in light.py) should all be created in `control4/__init__.py::async_setup_entry` and call `async_config_entry_first_refresh` on them?

nalin29 replied:
On reload the integration will work if the director is down during initialization and is brought up before next reload.

nalin29 replied:
I don't follow. If we fail to get an update here then the module has failed to setup as we are not able to get a request from controller and would require a reload like any other component. The coordinator is installed for each entity as the poll update mechanism. Failures there would be ignored. Only on setup is this an issue.

nalin29 replied:
This change can also be left for another pull request. I have one more change after this to address connection issues that is short and i am hoping we can merge this soon and discuss that in the shorter pull

emontnemery commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,406 @@
+"""Platform for Control4 Rooms."""
```
Can we come up with a more descriptive name than just "room"? Same comment for the media player sub class below.

nalin29 replied:
In control4 they are called rooms. The media control centers around room based audio so rooms is the correct description

emontnemery commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,406 @@
+    async_add_entities(entity_list, True)
+
+
+class Control4Room(Control4Entity, MediaPlayerEntity):
```
Please read up on the [entity naming rules](https://developers.home-assistant.io/docs/core/entity/#has_entity_name-true-mandatory-for-new-integrations), update as needed and then set `has_entity_name` to True. You may want to first do that for the existing light platform.

emontnemery commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,406 @@
+            coordinator,
+            name,
+            idx,
+            device_name=None,
+            device_manufacturer=None,
+            device_model=None,
```
This will set manufacturer, mode and name of the device to `None`, is that really what you want? Maybe set the device name to the name of the room and the name of the media player entity itself to `None` or something?

emontnemery commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,406 @@
+            entry_data,
+            coordinator,
+            name,
+            idx,
```
An index seems like a terrible device id, or is there a guarantee it will never ever change?
When creating the `Control4Room` instances, the corresponding variable is called `room_id`, is this in fact not an index but a unqie, never changing, ID of the room?

nalin29 replied:
Room id is the unique id for that room in control4 and the only unique identifier we have access to so we should use it for device id

emontnemery replied:
OK, but we should then call it `room_id`, not `idx`

nalin29 replied:
Done

emontnemery commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,406 @@
+
+    @property
+    def state(self):
+        """Return whether this room is on or off."""
```
I guess it's not a room which is on or off? 🤔

nalin29 replied:
In control 4 the room is technically marked as on or off of there are no media sources playing

emontnemery commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,406 @@
+
+        await self.coordinator.async_request_refresh()
+
+    def turn_on(self):
```
This is weird. How about setting the feature flag for `TURN_OFF` but not for `TURN_ON`, and after a call to `TURN_OFF` set the state to `MediaPlayerState.IDLE` or `MediaPlayerState.STANDBY` instead of `MediaPlayerState.OFF`?

nalin29 replied:
Yeah we can do that if we don't want to reflect the true state of off or on. Since in control4 by playing a source turns on the room and sets media

nalin29 replied:
Oh yeah. The UI breaks in many cases as we cannot start playing without the device being ON. Hence the necessity for this fake turn on.

nalin29 replied:
Done off sets state to idle no more soft on/fake off.

emontnemery commented:
Please open a documentation PR to make it clear what is supported: https://www.home-assistant.io/integrations/control4/

nalin29 commented:
> Please open a documentation PR to make it clear what is supported: https://www.home-assistant.io/integrations/control4/

https://github.com/home-assistant/home-assistant.io/pull/26196

granvillebarker commented:
Guys this control4 integration not getting merged has really been a pain for me.   I didn't realize that lawtancool had a much  newer version on his site, and implemented fan and climate based on this old outdated stuff in the current version of home assistant.   Then I rewrote it all to work with his latest integration on his git respository.  Now after reading all this, I don't know if nalin29 has updated lawtancools's repository, or if there are becoming many versions of this integration that are going in different directions, which is not good for anyone.   

The whole point of having all this is to prevent this type of stuff from happening, but when latest codes doesn't get merged its not good.   SO CAN ALL YOU GUYS GET OFF YOUR COUCHES AND GET THIS DONE!

nalin29 commented:
@granvillebarker 

I was waiting to try to get this merged beginning of year. Due to unfortunate circumstances the reviewer was unable to help get this merged so it sat stale hence my 3rd party. But I don't have continuous access to the control4 system anymore as I have moved. So any developments from my side will be slow. If you want this merged you can try and pick this up, it's very simple code and I can help if wanted.

emontnemery commented:
@nalin29 If you want to get the attention of the reviewer, you must hit the "ready for review" button. You replied to some of my review comments, but not all, are you planning to address the remaining ones?

nalin29 committed:
Merge branch 'home-assistant:dev' into dev

nalin29 committed:
update attribute names (#14)

nalin29 committed:
Merge branch 'home-assistant:dev' into dev

nalin29 commented:
hey @emontnemery . I just got access to the C4 system for a short time. Thought I could bang this revision out and test it. Can you review. There are some other changes needed to allow me to pass pylint due to changes with data coordinator default data type.

nalin29 committed:
Merge branch 'dev' into dev

emontnemery commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,407 @@
+    VIDEO = 2
+
+
+class _RoomSource:
+    """Room Source Data."""
+
+    def __init__(self, source_type: set[_SourceType], idx: int, name: str) -> None:
+        """Initialize Room Source Data."""
+        self.source_type = source_type
+        self.idx = idx
+        self.name = name
```
I would make this a dataclass, but that can be done in a follow-up PR
```suggestion
@dataclass
class _RoomSource:
    """Room Source Data."""

    source_type: set[_SourceType]
    idx: int
    name: str
```

nalin29 replied:
Done

emontnemery commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,407 @@
+    """Control4 Room entity."""
+
+    _attr_has_entity_name = True
+    _attr_name = None
```
This doesn't work, because it's overridden by the call to `super().__init__` below

nalin29 replied:
Done

emontnemery commented on homeassistant/components/control4/media_player.py:
```diff
@@ -0,0 +1,407 @@
+        self,
+        entry_data: dict,
+        coordinator: DataUpdateCoordinator[dict[int, dict[str, Any]]],
+        name: str,
```
I think you should pass `None` here

nalin29 replied:
we can' pass none here pylint won't allow. Is it an issue to provide the same name for enitity and device since there is only one entity I think that seems fine.

nalin29 replied:
Done

nalin29 committed:
change to data class and set off to idle (#15)

Co-authored-by: nalin29 <nalin29@github.com>

nalin29 committed:
Merge branch 'dev' into dev

nalin29 commented:
@emontnemery made all requested changes please take a look. 

If we are ready to merge i had to reopen the doc pull request as it was closed: https://github.com/home-assistant/home-assistant.io/pull/31910

Additionally, there is one more pull request I will open that will address stability since users are reporting to custom implementation that c4 api is unstable recently so i am adding retries to some of the class during setup.

emontnemery reviewed (approved):
Thanks, @nalin29 👍
